extensions/binding/org.eclipse.smarthome.binding.bosesoundtouch/src/main/java/org/eclipse/smarthome/binding/bosesoundtouch/handler/BoseSoundTouchHandler.java

```diff
@@ -410,15 +426,46 @@ private void checkConnection() {
         if (getThing().getStatus() != ThingStatus.ONLINE) {
             openConnection(); // try to reconnect....
         }
-        if (getThing().getStatus() == ThingStatus.ONLINE) {
+
+        if (getThing().getStatus() == ThingStatus.ONLINE && this.session != null && this.session.isOpen()) {
             try {
-                session.getRemote().sendString("HELLO");
-            } catch (IOException | NullPointerException e) {
+                this.session.getRemote().sendPing(null);
+                this.missedPongsCount.countDown();
+            } catch (IOException | NullPointerException | UnsupportedOperationException e) {
+                logger.warn("Sending websocket PING failed.");
                 onWebSocketError(e);
                 closeConnection();
                 openConnection();
             }
+        }
+    }
+    
+    /**
+     * A simple counter helper class 
+     */
+    private static class Counter {
+        private final int startValue;
+        private int currentValue;
+
+        Counter(int initialValue) {
+            this.startValue = initialValue;
+            this.currentValue = initialValue;
+        }
+
+        public synchronized void countDown() {
+            if (currentValue == 0) {
+                throw new UnsupportedOperationException("Cannot count down below 0!");

```

Comment:
`UnsupportedOperationException` seems an unsuitable choice here. `IllegalStateException` might be better, but imho simply removing this class and using an `int` as a counter and comparing it against the max would be much better readable.

Reply:
Totally agree.